Add parent navigation support for the navigator component

youknowriad opened the pull request:
Builds on top of https://github.com/WordPress/gutenberg/pull/47827 and addresses an issue raised in #47777 

## What and Why?

Parent or "back" navigation doesn't work as we expect it to work in Global Styles and Site editor. The reason is that `goBack` behaves like a browser history, it shows the previous screen that was visible. Sometimes navigation happens programmatically and sometimes the initial screen rendered is not the top level ones (like after reloading the page). For these situations, goBack doesn't work for us. What we really want is a child/parent relationship where the "back" button always goes back to the parent screen.

This PR adds support for that to the Navigator components.

## How?

There are two main ideas here. A user can define a "tree" of screens by relying on the paths: 

  - `/` is the root screen
  - `/something` is a child of root
  - `/:argument` is also a child of root
  - `/something/grandchild` is a child of `/something` if defined, otherwise it fallback to `/`

Basically to define the hierarchy you can split a path on the `/` character and move up until you find a screen that available.

The second idea of this PR is that `goToParent` is equivalent to `goToBack` and should restore the focus in the same way if the path is the previous in location history.

 - I've added a story to storybook to show how this works.

## Testing Instructions

 1- You can open the "nested navigator" story in storybook and check the behavior.


## ✍️ Dev Note

The `Navigator` component from the `@wordpress/components` package has been enhanced with two additional features:

- it can now match named arguments (ie. `/product/:productId`);
- it now offers a way to navigate to the parent screen via the `goToParent` function and the `NavigatorToParentButton` component.

These two features assume that `NavigatorScreen`s are assigned paths that are hierarchical and follow a URL-like scheme, where each path segment is separated by the `/` character.

youknowriad commented on packages/components/src/navigator/navigator-back-button/hook.ts:
```diff
@@ -17,18 +17,23 @@ export function useNavigatorBackButton(
 	const {
 		onClick,
 		as = Button,
+		goToParent: goToParentProp = false,
```
I've used a new prop in the "back" button to indicate whether it should behave like a "goBack" or "goToParent" by default. We could in theory build a another component but it felt too much.

ciampo replied:
Mhh, I see your point.

I just wonder if having them as separate buttons would be simpler to understand for the consumers of this component — we would have a 1:1 mapping between React components and `useNavigator` functions:

- `goTo` => `<NavigatorButton />`
- `goBack` => `<NavigatorBackButton />`
- `goToParent` => `<NavigatorToParentButton />` (example name)

youknowriad replied:
I don't care too much, I applied your suggestion.

ciampo commented:
Thank you for working on this so quickly!

For now, I focused my efforts on reviewing https://github.com/WordPress/gutenberg/pull/47827.

I should be able to take a look at this PR tomorrow.

youknowriad committed:
Add parent navigation support for the navigator component

youknowriad commented:
I've rebased this, it's still missing docs and changelog entry, but let's discuss implementation.

youknowriad commented on packages/components/src/navigator/navigator-provider/component.tsx:
```diff
@@ -66,7 +67,15 @@ function UnconnectedNavigatorProvider(
+	}, [ screens ] );
+	useEffect( () => {
+		currentLocationHistory.current = locationHistory;
+	}, [ locationHistory ] );
```
I hate to do this kind of useEffect. This is basically the perfect use case for the React's proposed useEvent. I'm looking forward to using it.

ciampo commented on packages/components/src/navigator/utils/router.ts:
```diff
@@ -21,3 +21,28 @@ export function patternMatch( path: string, screens: Screen[] ) {
 }
+
+export function findParent( path: string, screens: Screen[] ) {
+	if ( path[ 0 ] !== '/' ) {
```
For ease of reading, could we replace this with `startsWith` or with a Regex `/^\//.test( path )` ?

We should also add a unit test for when `path` does not start with `/`.

And finally, we should remember to document that, when using `goToParent`, the `path` needs to assume a URL-style format, with `/` as the folder separator

ciampo commented on packages/components/src/navigator/utils/router.ts:
```diff
@@ -21,3 +21,28 @@ export function patternMatch( path: string, screens: Screen[] ) {
+				const matchingFunction = match( screen.path, {
+					decode: decodeURIComponent,
+				} );
+				return matchingFunction( potentialParentPath ) !== false;
```
Since this is used to ensure the `potentialParentPath` has indeed a match with one of the screens, we should refactor the `matchingFunction` so that we are 100% ensured that it's same used by both `patternMatch()` and `findParent()`.

Otherwise in the future there would be the risk that the two "matching" functions get out of sync, and the expectation highlighted by this comment won't hold true.

ciampo commented on packages/components/src/navigator/utils/router.ts:
```diff
@@ -21,3 +21,28 @@ export function patternMatch( path: string, screens: Screen[] ) {
+	}
+	const pathParts = path.split( '/' );
+	let parentPath;
+	while ( pathParts.length > 1 && ! parentPath ) {
```
The fact that the `! parentPath` _implicitly_ checks for an empty string (on top of `undefined`) could bite us back in a a future refactor. Should we make the check more explicit, ie. 

```suggestion
	while ( pathParts.length > 1 && parentPath !== undefined && parentPath !== '' ) {
```

ciampo commented on packages/components/src/navigator/test/router.ts:
```diff
@@ -48,3 +48,51 @@ describe( 'patternMatch', () => {
+			{ id: 'route2', path: '/test' },
+		] );
+		expect( result ).toEqual( '/test' );
+	} );
```
A few other test cases that we could add:

```js
// No grandparent is found, fallback to `/`
findParent( '/test/nested/path', [
  { id: 'route1', path: '/other-path' },
  { id: 'route2', path: '/yet-another-path' },
  { id: 'root', path: '/' },
] );
```

```js
// No grandparent is found, and `/` is missing too 
findParent( '/test/nested/path', [
  { id: 'route1', path: '/other-path' },
  { id: 'route2', path: '/yet-another-path' },
] );
```

ciampo commented on packages/components/src/navigator/navigator-provider/component.tsx:
```diff
@@ -164,6 +177,34 @@ function UnconnectedNavigatorProvider(
+			if ( parentPath === undefined ) {
+				return;
+			}
+			const isBack =
+				currentLocationHistory.current.length > 1 &&
+				currentLocationHistory.current[
+					currentLocationHistory.current.length - 2
+				].path === parentPath;
+			if ( isBack ) {
+				goBack();
```
This is clever! I just wonder if it's a good idea to "mix and match" `goBack` and `goTo`, ie. chronological and hierarchical paradigms. What do you think?

Also, smaller nit: I would rename `isBack` to something more explicit (like `isSameAsPreviousPath`), to avoid the confusion with the `isBack` location option

youknowriad replied:
> This is clever! I just wonder if it's a good idea to "mix and match" goBack and goTo, ie. chronological and hierarchical paradigms. What do you think?

what would you do instead, we need to apply the exact same logic in this case.

ciampo replied:
I thought more about it and I think that it should ok the way it is in this PR (I in a [different convo](https://github.com/WordPress/gutenberg/pull/47883#discussion_r1101839766) I actually proposed to move this "short circuit"  directly to the `goTo` function, so that it can be leveraged also when calling `goTo( '/path', {isBack: true})`

ciampo commented on packages/components/src/navigator/types.ts:
```diff
@@ -12,11 +12,11 @@ export type MatchParams = Record< string, string | string[] >;
 
 type NavigateOptions = {
 	focusTargetSelector?: string;
+	isBack?: boolean;
```
By moving `isBack` to `NavigateOptions`, we're exposing it as an option to consumers of `Navigator` to pass `isBack` when calling `goTo`.

As of today, specifying `isBack === true` has two consequences:
- enables the usage of `focusTargetSelector` to restore focus
- animates backwards

By exposing `isBack` in `goTo`, my worry is that:
- consumers of the component would _technically_ be able to just use `goTo`, and completely disregard `goBack` and `goToParent`
- we'll need to be careful in the future with whatever implication we add to the `isBack` flag, in case it conflicts with `goTo` (which is a function "pushing" on top of the stack).

What are your thoughts on this one?

youknowriad replied:
Yeah I'm aware of it and I actually didn't see any harm. (It kind of enables the third option I proposed implicitly).  There's of course a possibility to hide it by having "two" go to functions. one public and one private but it felt unnecessary to me.

ciampo replied:
I definitely see the flexibility in keeping the `isBack` option available in `goTo` — as you say, consumers of `Navigator` could easily just use `goTo` and ignore `goBack` / `goToParent`.

On top of my previous worries, we should also consider that a consumer only using `goTo` would cause the `locationHistory` array to grow indefinitely, potentially causing a lot of memory usage.

I guess a potential compromise here could be:
- move the short circuit to `goBack()` from `goToParent` to `goTo`, so that we can call `goBack` wherever possible even when `goTo` is called
- potentially add a word of caution in the docs about memory usage?

youknowriad replied:
I always thought the location history should have a limit, so maybe instead we could put a max length to the history (like say 10 items)

youknowriad replied:
To be clear, I don't mind the other options, just sharing another one.

ciampo replied:
Alright, what about:

- we keep the `isBack` option available in `goTo`
- we move the "short circuit" to `goBack` from `goToParent` to `goTo`, so that we can internally use `goBack` as much as possible
- we add a max size to `locationHistory`. I think we could start with a high number (say 50?) and see how it goes? For the implementation, I was thinking: when the array is full, `push`ing one item will result in the array being `shift`ed, meaning that the oldest (first) item gets removed.

ciampo commented on packages/components/src/navigator/stories/index.tsx:
```diff
@@ -232,3 +232,64 @@ function ProductDetails() {
 	);
 }
+
+const NestedNavigatorTemplate: ComponentStory< typeof NavigatorProvider > = ( {
+	style,
+} ) => (
+	<NavigatorProvider
+		style={ { ...style, height: '100vh', maxHeight: '450px' } }
+		initialPath="/"
+	>
```
It's good practice for our stories to pass the `props` down to `NavigatorProvider`, instead of hardcoding them. This plays nicely with Storybook's controls too.

```suggestion
const NestedNavigatorTemplate: ComponentStory< typeof NavigatorProvider > = ( {
	style,
	...props
} ) => (
	<NavigatorProvider
		style={ { ...style, height: '100vh', maxHeight: '450px' } }
		{ ...props }
	>
```

At the bottom of the file, we can then add a `NestedNavigator.args` object with the `initialPath` prop.

I would actually argue that a better way to showcase the `goToParent` functionality is to have an `initialPath` different from the root:

```
NestedNavigator.args = {
	initialPath: '/child2/grandchild',
};
```

ciampo commented on packages/edit-site/src/components/sidebar-navigation-screen/index.js:
```diff
@@ -37,6 +37,7 @@ export default function SidebarNavigationScreen( {
 								__( 'Navigate to the previous view: %s' ),
 								parentTitle
 							) }
+							goToParent
```
What's the best way to test changes to this file?

youknowriad replied:
You can go to the navigation sidebar in the site editor and try navigating... that said, until #47777 there's only two level  of hierarchy in the tree meaning there's no difference between this branch and trunk so far. We'll have to merge this branch with #47777 to actually see the benefits.

ciampo reviewed (commented):
Thank you once more for working on this! I appreciate the thoughtfulness and the research of collaboration in the approach!

I gave a look at both Storybook and Global Styles, and the component seems to behave as expected 🎉 



Out of curiosity, [I checked current usages of Navigator outside of Gutenberg](https://wpdirectory.net/search/01GRVFW0MVZN0MYJAXTBG5SM56), and there's a bunch of plugins using it.

As discussed previously, the approach proposed in this PR seems the most reasonable in order to improve the component to fulfill product needs, without introducing breaking changes.

We'll need to make sure we update docs with clear explanations on how to use this component — mainly:
- using it chronologically (ie. using `goBack`) or hierarchically (`goToParent`)
- if using `goToParent`, `path` needs to have a URL-like format, starting with `/` and using `/` to separate each path segment.

Finally, I hope I'll be able to merge https://github.com/WordPress/gutenberg/pull/47910 soon (currently experiencing CI issues), in order to have the additional confidence boost given by extra unit tests

youknowriad committed:
Add new NavigatorToParentButton component

youknowriad committed:
Changes per review

youknowriad committed:
Add documentation

youknowriad commented:
> Out of curiosity, [I checked current usages of Navigator outside of Gutenberg](https://wpdirectory.net/search/01GRVFW0MVZN0MYJAXTBG5SM56), and there's a bunch of plugins using it.

I actually dove deeper into these results and it was interesting to note that all of them without exception use hierarchical paths. In other words, we could potentially decide to just replace goBack with goToParent and have no breakage.

youknowriad commented:
This PR should be ready for another round of review.

youknowriad committed:
Adapt e2e tests to new behavior

youknowriad commented on packages/edit-site/src/components/global-styles/ui.js:
```diff
@@ -267,11 +267,6 @@ function GlobalStylesStyleBook( { onClose } ) {
-				const depth = path.match( /\//g ).length;
-				for ( let i = 0; i < depth; i++ ) {
-					navigator.goBack();
-				}
```
@ciampo The kind of hack that this PR allows us to remove.

ciampo commented on packages/components/src/navigator/navigator-back-button/README.md:
```diff
@@ -10,22 +10,6 @@ The `NavigatorBackButton` component can be used to navigate to a screen and shou
-
 ### Inherited props
 
 `NavigatorBackButton` also inherits all of the [`Button` props](/packages/components/src/button/README.md#props), except for `href`.
```
Thank you for improving the docs, the `path` prop was wrong and the `onClick` is redundant if we mention the `Button` props.

Let's also mention the `target` prop

```suggestion
`NavigatorBackButton` also inherits all of the [`Button` props](/packages/components/src/button/README.md#props), except for `href` and `target`.
```

ciampo commented on packages/components/src/navigator/navigator-to-parent-button/README.md:
```diff
@@ -0,0 +1,15 @@
+
+### Inherited props
+
+`NavigatorBackButton` also inherits all of the [`Button` props](/packages/components/src/button/README.md#props), except for `href`.
```
```suggestion
`NavigatorToParentButton` also inherits all of the [`Button` props](/packages/components/src/button/README.md#props), except for `href` and `target`.
```

ciampo commented on packages/components/src/navigator/navigator-to-parent-button/component.tsx:
```diff
@@ -0,0 +1,65 @@
+ *   __experimentalNavigatorProvider as NavigatorProvider,
+ *   __experimentalNavigatorScreen as NavigatorScreen,
+ *   __experimentalNavigatorButton as NavigatorButton,
+ *   __experimentalNavigatorBackButton as NavigatorBackButton,
```
```suggestion
 *   __experimentalNavigatorToParentButton as NavigatorToParentButton,
```

ciampo commented on packages/components/src/navigator/types.ts:
```diff
@@ -57,8 +58,19 @@ export type NavigatorScreenProps = {
 	children: ReactNode;
 };
 
+export type NavigatorBackButtonHookProps = ButtonAsButtonProps & {
+	/**
+	 * Whether we should navigate to the parent screen.
+	 *
+	 * @default 'false'
+	 */
+	goToParent?: boolean;
+};
+
 export type NavigatorBackButtonProps = ButtonAsButtonProps;
 
+export type NavigatorToParentButtonProps = ButtonAsButtonProps;
```
Nitty: reshuffling types for better inheritance

```suggestion
export type NavigatorBackButtonProps = ButtonAsButtonProps;

export type NavigatorBackButtonHookProps = NavigatorBackButtonProps & {
	/**
	 * Whether we should navigate to the parent screen.
	 *
	 * @default 'false'
	 */
	goToParent?: boolean;
};

export type NavigatorToParentButtonProps = NavigatorBackButtonProps;
```

ciampo reviewed (commented):
> This PR should be ready for another round of review.

Tested in both Storybook and Site editor, and everything works as expected.

Apart from the inline comments, before merging we should also:

1. Update the `navigator` object docs

<details>
<summary>Suggested changes</summary>

````diff
diff --git a/packages/components/src/navigator/navigator-provider/README.md b/packages/components/src/navigator/navigator-provider/README.md
index 3afd84b559..47c2af4b9d 100644
--- a/packages/components/src/navigator/navigator-provider/README.md
+++ b/packages/components/src/navigator/navigator-provider/README.md
@@ -34,13 +34,17 @@ const MyNavigation = () => (
   </NavigatorProvider>
 );
 ```
+
 **Important note**
 
-Parent/child navigation only works if the path you define are hierarchical. For example:
- - `/` is the root of all paths. 
- - `/parent/child` is a child of `/parent`. 
- - `/parent/child/grand-child` is a child of `/parent/child`.
- - `/parent/:param` is a child of `/parent` as well.
+Parent/child navigation only works if the path you define are hierarchical, following a URL-like scheme where each path segment is separated by the `/` character.
+
+For example:
+
+- `/` is the root of all paths. There should always be a screen with `path="/"`.
+- `/parent/child` is a child of `/parent`.
+- `/parent/child/grand-child` is a child of `/parent/child`.
+- `/parent/:param` is a child of `/parent` as well.
 
 ## Props
 
@@ -65,6 +69,15 @@ The `goTo` function allows navigating to a given path. The second argument can a
 The available options are:
 
 - `focusTargetSelector`: `string`. An optional property used to specify the CSS selector used to restore focus on the matching element when navigating back.
+- `isBack`: `boolean`. An optional property used to specify whether the navigation should be considered as backwards (thus enabling focus restoration when possible, and causing the animation to be backwards too)
+
+### `goToParent`: `() => void;`
+
+The `goToParent` function allows navigating to the parent screen.
+
+Parent/child navigation only works if the path you define are hierarchical (see note above).
+
+When a match is not found, the function will try to recursively navigate the path hierarchy until a matching screen (or the root `/`) are found.
 
 ### `goBack`: `() => void`
````
</details>

2. Ideally, add tests to `packages/components/src/navigator/test/index.tsx` using the `NavigatorToParentButton` component

> I actually dove deeper into these results and it was interesting to note that all of them without exception use hierarchical paths. In other words, we could potentially decide to just replace goBack with goToParent and have no breakage.

That's an interesting reflection — it means that, until now, all consumers of `Navigator` are basically using it to browse a tree-like structure of screens by transition from parent to child, or from child to parent.

I still think that offering chronological `goBack` could prove useful for a consumer that doesn't necessarily navigate screens from parent to child (or from child to parent).

For example, imagine these scenarios 
- we show the top menu (typography, blocks, ...) in a navigation bar that is always visible in every screen. A user could use that navigation bar "jump" across screens.
- a user is editing typography settings for a specific block. We could show a "go to global typography settings" button that takes them to the global typography button

In both scenarios, a "go to parent" button (i.e. go to home) and a "back" button (ie., back to whatever screen was visited previously) would both be useful for different reasons.

youknowriad committed:
Improve history behavior

youknowriad committed:
Fix button docs

Co-authored-by: Marco Ciampini <marco.ciampo@gmail.com>

youknowriad committed:
Fix button docs

Co-authored-by: Marco Ciampini <marco.ciampo@gmail.com>

youknowriad committed:
fix import doc

Co-authored-by: Marco Ciampini <marco.ciampo@gmail.com>

youknowriad committed:
Shuffling types

Co-authored-by: Marco Ciampini <marco.ciampo@gmail.com>

youknowriad committed:
Improve docs

ciampo commented on packages/components/src/navigator/navigator-provider/component.tsx:
```diff
@@ -132,7 +175,12 @@ function UnconnectedNavigatorProvider(
 				return [
-					...prevLocationHistory.slice( 0, -1 ),
+					...prevLocationHistory.slice(
+						prevLocationHistory.length > MAX_HISTORY_LENGTH - 2
```
The current check will enforce an effective max length of `MAX_HISTORY_LENGTH - 1`. To enforce a max length of `MAX_HISTORY_LENGTH` we should change it to

```suggestion
						prevLocationHistory.length > MAX_HISTORY_LENGTH - 1
```

youknowriad committed:
Add unit test

youknowriad commented:
I think I've addressed all the feedback and applied all suggestions @ciampo 

That said, CI and types started failing randomly, not sure exactly what I did that triggered that? Maybe your eagle eye can spot anything here?

youknowriad committed:
Fix type failures

youknowriad commented:
Ok the type failures were due to the import of the button component in the test. It should be fixed now.

ciampo commented on packages/components/src/navigator/test/index.tsx:
```diff
@@ -13,11 +13,13 @@ import { useState } from '@wordpress/element';
 /**
  * Internal dependencies
  */
+import { Button } from '../../button';
```
A more standard way of importing the `Button` in the package is

```suggestion
import Button from '../../button';
```

ciampo reviewed (approved):
Left two more minor comments, but we can go ahead and merge once they are solved
🚀 

Thank you for the collaboration on this one! `Navigator` feels in a much better place.

youknowriad committed:
Update button import

Co-authored-by: Marco Ciampini <marco.ciampo@gmail.com>

youknowriad committed:
Small fix

Co-authored-by: Marco Ciampini <marco.ciampo@gmail.com>

ntsekouras commented:
Cherry-picked this PR to the wp/6.2 branch.
